### PR TITLE
KN-13371 Exception on loading localization from cache

### PR DESCRIPTION
**Pull request by AlexPhilippov**

### CODE REVIEW

File: Core/LocalizationManager.cs [hunk 1]
```diff
@@ -157,71 +157,56 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
-				for (int i = 0; i < _initialBundlesCount; i++)
-					LocalizationCache.Insert(GetBundleName(culture, i), bundles[i]);
-			}
+		private LocalizedObjectList UpdateSingleBundle(CultureInfo culture, string bundleHashStart, IEnumerable<ILocalizedObject> allObjects)
+		{
+			var bundleObjects = allObjects.Where(x => x.Hash.ToLowerInvariant().Substring(0, 2) == bundleHashStart);
+			//only objects with correct Hash are loaded
+			var newBundle = new LocalizedObjectList();
+			foreach (var obj in bundleObjects)
+				newBundle.Add(obj);
```
Comment (pstarkov) [inline Core/LocalizationManager.cs]: AddRange

File: Core/LocalizationManager.cs [hunk 2]
```diff
@@ -157,71 +157,56 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
+			LocalizationCache.Insert(GetBundleName(culture, bundleHashStart), newBundle);
+			return newBundle;
+		}
 
-			return result;
+		private static string GetBundleName(CultureInfo culture, string bundleHashStart)
```
Comment (pstarkov) [inline Core/LocalizationManager.cs]: rename bundleHashStart to bundleIndex

File: Core/LocalizationManager.cs [hunk 3]
```diff
@@ -157,71 +157,56 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
+		{
+			return string.Format("{0}_bundle{1}", culture.Name, bundleHashStart);
 		}
 
-		private static string GetBundleName(CultureInfo culture, int bundleIndex)
+		private static string BundleIndexToHashStart(int bundleIndex)
```
Comment (pstarkov) [inline Core/LocalizationManager.cs]: BundleIndexHex

File: Core/LocalizationManager.cs [hunk 4]
```diff
@@ -157,71 +157,56 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
 		{
-			return string.Format("{0}_bundle{1}", culture.Name, bundleIndex);
+			return bundleIndex.ToString("x2");
 		}
 
-		private int GetBundleIndex(string hash)
+		private static string GetBundleHashStart(string hash)
```
Comment (pstarkov) [inline Core/LocalizationManager.cs]: GetBundleHex

File: Core/LocalizationManager.cs (hunk 1 above)
Comment (AlexPhilippov) [inline Core/LocalizationManager.cs]: That's Dictionary-derived class, there is no such method

File: Core/LocalizationManager.cs [hunk 5]
```diff
@@ -157,71 +157,56 @@ public IEnumerable<Object> GetScriptResources(CultureInfo culture)
...
 
-				for (int i = 0; i < _initialBundlesCount; i++)
-					LocalizationCache.Insert(GetBundleName(culture, i), bundles[i]);
-			}
+		private LocalizedObjectList UpdateSingleBundle(CultureInfo culture, string bundleHashStart, IEnumerable<ILocalizedObject> allObjects)
+		{
+			var bundleObjects = allObjects.Where(x => x.Hash.ToLowerInvariant().Substring(0, 2) == bundleHashStart);
+			//only objects with correct Hash are loaded
+			var newBundle = new LocalizedObjectList();
+			foreach (var obj in bundleObjects)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: add line

File: Core/LocalizationManager.cs [hunk 6]
```diff
@@ -232,16 +217,17 @@ public IEnumerable<CultureInfo> GetCultures()
...
 		{
-			var bundlesToDelete = new HashSet<string>();
+			var bundlesToUpdate = new HashSet<string>();
 			foreach (var obj in list)
-				bundlesToDelete.Add(GetBundleName(culture, GetBundleIndex(obj.Hash)));
-			foreach (var bundleName in bundlesToDelete)
-				LocalizationCache.Remove(bundleName);
+				bundlesToUpdate.Add(GetBundleHashStart(obj.Hash));
+			var allObjects = Repository.GetAll(culture);
+			foreach (var bundle in bundlesToUpdate)
```
Comment (polyakova) [inline Core/LocalizationManager.cs]: add line

**Pull request merged**